geoapps/io/validators.py

```diff
@@ -0,0 +1,244 @@
+#  Copyright (c) 2021 Mira Geoscience Ltd.
+#
+#  This file is part of geoapps.
+#
+#  geoapps is distributed under the terms and conditions of the MIT License
+#  (see LICENSE file at the root of this source code package).
+
+from typing import Any, Dict, List, Tuple
+
+import numpy as np
+
+from .constants import required_parameters, validations
+
+
+class InputValidator:
+    """
+    Validations for inversion parameters.
+
+    Attributes
+    ----------
+    input : Input file contents parsed to dict.
+
+    Methods
+    -------
+    validate_input(input)
+        Validates input params and contents/type/shape/keys of values.
+    validate(param value)
+        Validates parameter values, types, shapes, and keys.
+
+    """
+
+    def __init__(self, input: Dict[str, Any] = None):
+        self.input = input
+        self.current_validation = None
+
+    @property
+    def input(self):
+        return self._input
+
+    @input.setter
+    def input(self, val):
+        if val is not None:
+            self.validate_input(val)
+        self._input = val
+
+    def validate_input(self, input: Dict[str, Any]) -> None:
+        """
+        Validates input params and contents/type/shape/requirements of values.
+
+        Calls validate method on individual key/value pairs in input, and
+        handles validations requiring knowledge of other parameters.
+
+        Parameters
+        ----------
+        input : Input file contents parsed to dict.
+
+        Raises
+        ------
+        ValueError, TypeError KeyError whenever an input parameter fails one of
+        it's value/type/shape/requirement validations.
+        """
+
+        self._validate_required_parameters(input)
+
+        if "data" in input.keys():
+            require_keys = "type" not in input["data"].keys()
+            require_keys |= "name" not in input["data"].keys()
+            if require_keys:
+                raise KeyError("Data 'type' and 'name' must not be empty.")
+        elif "reference_model" in input.keys():
+            if "value" in input["reference_model"].keys():
```

Comment:
Same here, I think we want to generalize `reference_model` and assume that only a vector, float or None is provided. To simplify this let's split the `vector` case into 3 distinct reference models (amplitude, inclination, declination). Same would technically apply to starting model also... just not implemented here.